src/lib/math/bigint/bigint.h

```diff
@@ -649,41 +656,53 @@ class BOTAN_PUBLIC_API(2, 0) BigInt final {
        * Store BigInt-value in a given byte array
        * @param buf destination byte array for the integer value
        */
-      void binary_encode(uint8_t buf[]) const;
+      void binary_encode(uint8_t buf[]) const {
+         // assumes that `buf` points to a memory region with enough bytes
+         binary_encode(std::span{buf, bytes()});
+      }
 
       /**
-       * Store BigInt-value in a given byte array. If len is less than
-       * the size of the value, then it will be truncated. If len is
-       * greater than the size of the value, it will be zero-padded.
-       * If len exactly equals this->bytes(), this function behaves identically
-       * to binary_encode.
-       *
+       * Store BigInt-value in a given byte array
        * @param buf destination byte array for the integer value
        * @param len how many bytes to write
        */
-      void binary_encode(uint8_t buf[], size_t len) const;
+      void binary_encode(uint8_t buf[], size_t len) const { binary_encode(std::span{buf, len}); }
+
+      /**
+       * Store BigInt-value in a given byte array. If the buffer length
+       * is less than the size of the value, then it will be truncated.
+       * If it is greater than the size of the value, it will be zero-padded.
+       * If the buffer length exactly equals this->bytes(), this function behaves
+       * identically to binary_encode.
+       *
+       * @param buf destination byte array for the integer value
+       */
+      void binary_encode(std::span<uint8_t> buf) const;
 
       /**
        * Read integer value from a byte array with given size
        * @param buf byte array buffer containing the integer
        * @param length size of buf
        */
-      void binary_decode(const uint8_t buf[], size_t length);
+      void binary_decode(const uint8_t buf[], size_t length) { binary_decode(std::span{buf, length}); }
 
       /**
        * Read integer value from a byte vector
        * @param buf the vector to load from
        */
-      template <typename Alloc>
-      void binary_decode(const std::vector<uint8_t, Alloc>& buf) {
-         binary_decode(buf.data(), buf.size());
-      }
+      void binary_decode(std::span<const uint8_t> buf);
 
       /**
-       * Place the value into out, zero-padding up to size words
-       * Throw if *this cannot be represented in size words
+       * Place the value into @p out, zero-padding up to @p size words
+       * Throw if *this cannot be represented in @p size words
        */
-      void encode_words(word out[], size_t size) const;
+      void encode_words(word out[], size_t size) const { encode_words(std::span{out, size}); }
```

Comment:
Should we deprecate some of these ptr+len interfaces?

Reply:
Yes please! :)